source/common/upstream/health_discovery_service.cc

```diff
@@ -298,24 +362,166 @@ HdsCluster::HdsCluster(Server::Admin& admin, Runtime::Loader& runtime,
+                       api);
+  }
+}
+
+void HdsCluster::updateHealthchecks(
+    const Protobuf::RepeatedPtrField<envoy::config::core::v3::HealthCheck>& health_checks,
+    AccessLog::AccessLogManager& access_log_manager, Runtime::Loader& runtime,
+    Random::RandomGenerator& random, Event::Dispatcher& dispatcher, Api::Api& api) {
+  std::vector<Upstream::HealthCheckerSharedPtr> health_checkers;
+  HealthCheckerMap health_checkers_map;
+
+  for (auto& health_check : health_checks) {
+    // Check to see if this exact same health_check config already has a health checker.
+    auto health_checker = health_checkers_map_.find(health_check);
+    if (health_checker != health_checkers_map_.end()) {
+      // If it does, use it.
+      health_checkers_map.insert({health_check, health_checker->second});
+      health_checkers.push_back(health_checker->second);
+    } else {
+      // If it does not, create a new one.
+      auto new_health_checker =
+          Upstream::HealthCheckerFactory::create(health_check, *this, runtime, random, dispatcher,
+                                                 access_log_manager, validation_visitor_, api);
+      health_checkers_map.insert({health_check, new_health_checker});
+      health_checkers.push_back(new_health_checker);
+
+      // Start these health checks now because upstream assumes they already have been started.
+      new_health_checker->start();
+    }
+  }
+
+  // replace our member data structures with our newly created ones.
+  health_checkers_ = std::move(health_checkers);
+  health_checkers_map_ = std::move(health_checkers_map);
+}
+void HdsCluster::updateHosts(
+    const Protobuf::RepeatedPtrField<envoy::config::endpoint::v3::LocalityLbEndpoints>&
+        locality_endpoints,
+    bool update_socket_matches) {
+  // Create the data structures needed for PrioritySet::update.
+  HostVectorSharedPtr hosts = std::make_shared<std::vector<HostSharedPtr>>();
+  std::vector<HostSharedPtr> hosts_added;
+  std::vector<HostSharedPtr> hosts_removed;
+  std::vector<HostVector> hosts_by_locality;
+
+  // Use for delta update comparison.
+  HostsMap hosts_map;
+
+  for (auto& endpoints : locality_endpoints) {
+    hosts_by_locality.emplace_back();
+    for (auto& endpoint : endpoints.lb_endpoints()) {
+      LocalityEndpointTuple endpoint_key = {endpoints.locality(), endpoint};
+
+      // Check to see if this exact Locality+Endpoint has been seen before.
+      auto host_pair = hosts_map_.find(endpoint_key);
+      HostSharedPtr host;
+      if (host_pair != hosts_map_.end()) {
+        // If we have this exact pair, save the shared pointer.
+        host = host_pair->second;
+
+        // If our socket_matcher changed, update this host with the new info.
+        if (update_socket_matches) {
+          // We can guarantee this is a HostImpl, so we call its member function.
+          static_cast<HostImpl&>(*host).updateClusterInfo(info_);
```

Comment:
This seems kind of awful. Is there a way to avoid the cast? If not, can we at least use a dynamic down cast?

Reply:
Oops. Yeah, a dynamic_cast would be much better.

In terms of avoiding the cast, I tried very hard. The difficult part is that the HostImpl inherits from both Host and HostDescriptionImpl, and ClusterInfo exists as a member of HostDescriptionImpl. The only way to access this member is with a cast AFAIK, especially since there is no relation from the parent Host class and HostDescriptionImpl. Unless there is some pattern for this that I couldn't think of.